# Data sources

This page lists all data source modules officially supported by BioDWH2. Details for each module are available by clicking on the respective ID. As we strive to include each data source as a whole, some modules are usable, but still marked as work on progress as some information are not yet included or the mapping layer isn't finished yet. If you're interested in a specific data source, please open a [discussion](https://github.com/BioDWH2/BioDWH2/discussions/categories/data-source-requests).

> ⚠️ Please be aware: Databases available online may always be subject to change. Although we try to adjust our data source modules as quickly as possible there is always the possibility of a module not working correctly! If you experience any problems, please report it using the [GitHub issues](https://github.com/BioDWH2/BioDWH2/issues). This helps us fix issues quicker and all users to have a nice experience.

## Usable

| ID                            | Name                                                                                                                                                    | License                                                               | Website                                                               |
|-------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------|-----------------------------------------------------------------------|
| ADReCS                        | Adverse Drug Reaction Classification System                                                                                                             | [CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) | [Link](https://bioinf.xmu.edu.cn/ADReCS/)                             |
| BasicFormalOntology           | Basic Formal Ontology                                                                                                                                   | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://basic-formal-ontology.org)                             |
| BiologicalSpatialOntology     | Biological Spatial Ontology                                                                                                                             | [CC BY 3.0](https://creativecommons.org/licenses/by/3.0/)             | [Link](https://github.com/obophenotype/biological-spatial-ontology)   |
| BioM2MetDisease               | BioM2MetDisease                                                                                                                                         |                                                                       | [Link](http://bio-bigdata.hrbmu.edu.cn/BioM2MetDisease/)              |
| BRENDATissueOntology          | BRENDA Tissue Ontology                                                                                                                                  | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://www.brenda-enzymes.org/ontology.php?ontology_id=3)     |
| CancerDrugsDB                 | A curated listing of licensed cancer drugs by the Anticancer Fund                                                                                       |                                                                       | [Link](https://www.anticancerfund.org/en/cancerdrugs-db)              |
| CellOntology                  | Cell Ontology                                                                                                                                           | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://github.com/obophenotype/cell-ontology)                 |
| CMAUP                         | Collective Molecular Activities of Useful Plants                                                                                                        |                                                                       | [Link](http://bidd.group/CMAUP/)                                      |
| DiseaseOntology               | Human Disease Ontology                                                                                                                                  | CC0 1.0                                                               | [Link](https://disease-ontology.org)                                  |
| DrugBank                      | DrugBank                                                                                                                                                | [CC BY-NC 4.0](https://creativecommons.org/licenses/by-nc/4.0/)       | [Link](http://drugbank.ca)                                            |
| DrugCentral                   | DrugCentral                                                                                                                                             | [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)       | [Link](http://drugcentral.org)                                        |
| EFO                           | Experimental Factor Ontology                                                                                                                            | Apache-2.0                                                            | [Link](https://www.ebi.ac.uk/efo/)                                    |
| EMA                           | European Medicines Agency's (EMA) medicine-related data [EPAR and HMPC]                                                                                 |                                                                       | [Link](https://www.ema.europa.eu/en/medicines/download-medicine-data) |
| ENZYME                        | Expasy ENZYME                                                                                                                                           | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://enzyme.expasy.org)                                     |
| Gene2Phenotype                | Datasets for use in diagnostic variant filtering                                                                                                        |                                                                       | [Link](https://www.ebi.ac.uk/gene2phenotype)                          |
| GeneOntology                  | Gene Ontology                                                                                                                                           | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](http://geneontology.org)                                       |
| HGNC                          | HUGO Gene Nomenclature Committee                                                                                                                        | CC0 1.0                                                               | [Link](https://www.genenames.org)                                     |
| HPO                           | Human Phenotype Ontology                                                                                                                                | HPO license                                                           | [Link](https://hpo.jax.org)                                           |
| HPRD                          | Human Protein Reference Database                                                                                                                        |                                                                       | [Link](http://www.hprd.org)                                           |
| InterPro                      | InterPro                                                                                                                                                |                                                                       | [Link](https://www.ebi.ac.uk/interpro/)                               |
| MammalianPhenotypeOntology    | Mammalian Phenotype Ontology                                                                                                                            | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://github.com/mgijax/mammalian-phenotype-ontology)        |
| MassSpectrometryOntology      | Mass Spectrometry Ontology                                                                                                                              | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://github.com/HUPO-PSI/psi-ms-CV)                         |
| MED-RT                        | Medication Reference Terminology (MED-RT™)                                                                                                              | UMLS license                                                          | [Link](https://evs.nci.nih.gov/ftp1/MED-RT/)                          |
| miR2Disease                   | miR2Disease                                                                                                                                             |                                                                       | [Link](http://www.mir2disease.org)                                    |
| miRBase                       | miRBase                                                                                                                                                 | CC0 with attribution                                                  | [Link](https://www.mirbase.org)                                       |
| miRDB                         | miRDB                                                                                                                                                   |                                                                       | [Link](https://www.mirdb.org)                                         |
| miRTarBase                    | miRTarBase                                                                                                                                              |                                                                       | [Link](https://mirtarbase.cuhk.edu.cn)                                |
| MolecularInteractionsOntology | Molecular Interactions Controlled Vocabulary                                                                                                            | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://github.com/HUPO-PSI/psi-mi-CV)                         |
| Mondo                         | Mondo Disease Ontology                                                                                                                                  | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://mondo.monarchinitiative.org)                           |
| NDF-RT                        | National Drug File – Reference Terminology (NDF-RT™)                                                                                                    | UMLS license                                                          | [Link](https://evs.nci.nih.gov/ftp1/NDF-RT/)                          |
| Negatome                      | Negatome                                                                                                                                                |                                                                       | [Link](http://mips.helmholtz-muenchen.de/proj/ppi/negatome/)          |
| NeuroBehaviourOntology        | Neuro Behaviour Ontology                                                                                                                                |                                                                       | [Link](https://github.com/obo-behavior/behavior-ontology)             |
| OMIM                          | Online Mendelian Inheritance in Man®                                                                                                                    | OMIM license                                                          | [Link](https://www.omim.org)                                          |
| PharmGKB                      | PharmGKB                                                                                                                                                | [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)       | [Link](https://www.pharmgkb.org)                                      |
| PhenotypeAndTraitOntology     | Phenotype And Trait Ontology                                                                                                                            | [CC BY 3.0](https://creativecommons.org/licenses/by/3.0/)             | [Link](https://github.com/pato-ontology/pato)                         |
| ReDO-DB                       | A curated listing of non-cancer drugs which have shown some evidence of anticancer activity by the Anticancer Fund                                      |                                                                       | [Link](https://www.anticancerfund.org/en/redo-db)                     |
| ReDOTrialsDB                  | A curated database, produced by the Anticancer Fund, of active clinical trials investigating the use of non-cancer drugs as potential cancer treatments |                                                                       | [Link](https://www.anticancerfund.org/en/redo-trials-db)              |
| RNALocate                     | RNA subcellular localization database                                                                                                                   |                                                                       | [Link](http://www.rna-society.org/rnalocate/)                         |
| SequenceOntology              | Sequence Ontology                                                                                                                                       | [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)       | [Link](http://www.sequenceontology.org)                               |
| SIDER                         | SIDER: Side Effect Resource                                                                                                                             | [CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) | [Link](http://sideeffects.embl.de)                                    |
| TRRUST                        | TRRUST                                                                                                                                                  | [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)       | [Link](https://www.grnpedia.org/trrust/)                              |
| UberonOntology                | Uberon Ontology                                                                                                                                         | [CC BY 3.0](https://creativecommons.org/licenses/by/3.0/)             | [Link](https://github.com/obophenotype/uberon)                        |
| UNII                          | FDA Substance Registration System - Unique Ingredient Identifier                                                                                        |                                                                       | [Link](https://fdasis.nlm.nih.gov/srs/)                               |
| UnitOntology                  | Unit Ontology                                                                                                                                           | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://github.com/bio-ontology-research-group/unit-ontology)  |
| USDA-PLANTS                   | USDA PLANTS Database                                                                                                                                    | Free with attribution                                                 | [Link](https://plants.sc.egov.usda.gov)                               |

## Work in Progress

| ID                   | Name                                                                               | License                                                               | Website                                                                                                  |
|----------------------|------------------------------------------------------------------------------------|-----------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------|
| BRENDA               | BRENDA                                                                             | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://www.brenda-enzymes.org)                                                                   |
| CanadianNutrientFile | CanadianNutrientFile                                                               |                                                                       | [Link](https://www.canada.ca/en/health-canada/services/food-nutrition/healthy-eating/nutrient-data.html) |
| ClinicalTrials       | ClinicalTrials.gov Clinical Trials Database                                        |                                                                       | [Link](https://ClinicalTrials.gov)                                                                       |
| CPDB                 |                                                                                    |                                                                       | [Link](http://cpdb.molgen.mpg.de)                                                                        |
| DGIdb                | The Drug Gene Interaction Database                                                 |                                                                       | [Link](http://dgidb.org)                                                                                 |
| DISEASES             | Gene-Disease association database                                                  | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://diseases.jensenlab.org)                                                                   |
| GenCC                | GenCC                                                                              | CC0 1.0                                                               | [Link](https://thegencc.org)                                                                             |
| GuideToPharmacology  | IUPHAR - GuideToPharmacology                                                       | [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)       | [Link](https://www.guidetopharmacology.org)                                                              |
| GWASCatalog          | GWAS Catalog                                                                       |                                                                       | [Link](https://www.ebi.ac.uk/gwas/home)                                                                  |
| HAMAP                | Expasy HAMAP                                                                       |                                                                       | [Link](https://hamap.expasy.org)                                                                         |
| HERB                 | Experiment- and reference-guided database of traditional Chinese medicine          |                                                                       | [Link](http://herb.ac.cn)                                                                                |
| HMDB                 | Human Metabolome Database                                                          |                                                                       | [Link](https://hmdb.ca)                                                                                  |
| IntAct               | IntAct                                                                             | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://www.ebi.ac.uk/intact/)                                                                    |
| ITIS                 | Integrated Taxonomic Information System                                            | CC0 with attribution                                                  | [Link](https://www.itis.gov)                                                                             |
| KEGG                 | Kyoto Encyclopedia of Genes and Genomes                                            |                                                                       | [Link](https://www.kegg.jp)                                                                              |
| PathwayCommons       | Data warehouse of different pathway databases                                      |                                                                       | [Link](http://www.pathwaycommons.org)                                                                    |
| PROSITE              | Expasy PROSITE & ProRule                                                           | [CC BY-NC-ND 4.0](https://creativecommons.org/licenses/by-nc-nd/4.0/) | [Link](https://prosite.expasy.org)                                                                       |
| RefSeq               | RefSeq                                                                             |                                                                       | [Link](https://www.ncbi.nlm.nih.gov/refseq/)                                                             |
| RNADisease           | RNADisease                                                                         | non-commercial with attribution                                       | [Link](http://www.rnadisease.org)                                                                        |
| RNAInter             | RNAInter                                                                           | non-commercial with attribution                                       | [Link](http://www.rnainter.org)                                                                          |
| STITCH               | STITCH Chemical-Protein Interaction Networks                                       |                                                                       | [Link](http://stitch.embl.de)                                                                            |
| STRING               | STRING Protein-Protein Interaction Networks                                        |                                                                       | [Link](https://string-db.org)                                                                            |
| T3DB                 | T3DB                                                                               |                                                                       | [Link](http://www.t3db.ca)                                                                               |
| TarBase              | TarBase                                                                            |                                                                       | [Link](http://microrna.gr/tarbase)                                                                       |
| TISSUES              | Tissue expression database                                                         | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://tissues.jensenlab.org)                                                                    |
| TTD                  | Therapeutic Target Database (TTD)                                                  |                                                                       | [Link](http://db.idrblab.net/ttd/)                                                                       |
| UniProt              | Functional information on proteins, with accurate, consistent and rich annotation. | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)             | [Link](https://www.uniprot.org)                                                                          |
| OpenTargets          | Systematic drug target identification and prioritisation.                          |                                                                       | [Link](https://www.opentargets.org)                                                                      |

## Data Source Details

----

### ADReCS

Adverse Drug Reaction Classification System (ADReCS) is a comprehensive ADR ontology database that provides both standardization and hierarchical classification of ADR terms. The ADReCS integrates ADR and drug information collected from public medical repositories like FAERS, DailyMed, SIDER4.1, MedDRA, DrugBank, PubChem, and so on.

- **Website**: [https://bioinf.xmu.edu.cn/ADReCS/](https://bioinf.xmu.edu.cn/ADReCS/)
- **License**: unknown
- **Citation**: ```Mei-Chun Cai, Quan Xu, Yan-Jing Pan, Wen Pan, Nan Ji, Yin-Bo Li, Hai-Jing Jin, Ke Liu and Zhi-Liang Ji. (2015) ADReCS: an ontology database for aiding standardization and hierarchical classification of adverse drug reaction terms. Nucl. Acids Res. 43 (D1): D907-D913. doi: [10.1093/nar/gku1066](https://dx.doi.org/10.1093/nar/gku1066). Epub 2014 Oct 31.```
- **Files used**:
  - `DRUG_ADR.zip`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### BasicFormalOntology

The Basic Formal Ontology (BFO)

- **Website**: [https://basic-formal-ontology.org](https://basic-formal-ontology.org)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**:
- **Files used**:
  - At the moment, the ontology is downloaded in OBO format (```bfo.obo```).

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### BiologicalSpatialOntology

- **Website**: [https://github.com/obophenotype/biological-spatial-ontology](https://github.com/obophenotype/biological-spatial-ontology)
- **License**: [CC BY 3.0](https://creativecommons.org/licenses/by/3.0/)
- **Citation**:
- **Files used**:
  - At the moment, the ontology is downloaded in OBO format (```bspo-full.obo```).

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### BioM2MetDisease

- **Website**: [http://bio-bigdata.hrbmu.edu.cn/BioM2MetDisease/](http://bio-bigdata.hrbmu.edu.cn/BioM2MetDisease/)
- **License**:
- **Citation**:
- **Files used**:
  - ```BioM2MetDiseaseData.txt```

----

### BRENDA

- **Website**: [https://www.brenda-enzymes.org](https://www.brenda-enzymes.org)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**: ```Chang A, Jeske L, Ulbrich S, et al. BRENDA, the ELIXIR core data resource in 2021: new developments and updates. Nucleic Acids Research. 2021;49(D1):D498-D508. doi:``` [10.1093/nar/gkaa1025](http://doi.org/10.1093/nar/gkaa1025)
- **Files used**:
  - `brenda_[...].json.tar.gz`

**Configuration properties**:

| Property        | Values     | Description                                                      |
|-----------------|------------|------------------------------------------------------------------|
| forceExport     | true/false | Force graph export even if nothing changed                       |
| licenseAccepted | true/false | Manual check if the license of BRENDA has been read and accepted |

----

### BRENDATissueOntology

BRENDA Tissue Ontology (BTO). A structured controlled vocabulary for the source of an enzyme comprising tissues,
cell lines, cell types and cell cultures.

- **Website**: [https://www.brenda-enzymes.org/ontology.php?ontology_id=3](https://www.brenda-enzymes.org/ontology.php?ontology_id=3)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**:
- **Files used**:
  - At the moment, the ontology is downloaded in OBO format (```bto.obo```).

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### CanadianNutrientFile

Health Canada publishes databases which list nutrient values in Canadian foods. The Canadian Nutrient File (CNF) is a comprehensive, computerized bilingual database that reports up to 152 nutrients in over 5690 foods. The database can help you find values for nutrients such as vitamins, minerals, protein, energy, fat and many more, and is updated periodically.

- **Particularities**:
  - Some files are broken and contain empty lines, lines with only ',' and some lines are split into multiple lines. This is handled and fixed when importing the files.
  - Sometimes a links are wrong and point to not existing entries. This concerns measurements, nutrients and food source. This broken links will be ignored.
- **Website**: [https://www.canada.ca/en/health-canada/services/food-nutrition/healthy-eating/nutrient-data.html](https://www.canada.ca/en/health-canada/services/food-nutrition/healthy-eating/nutrient-data.html)
- **License**: [copyright-guidelines-canadian-nutrient-file](https://www.canada.ca/en/health-canada/services/food-nutrition/healthy-eating/nutrient-data/copyright-guidelines-canadian-nutrient-file.html)
- **Citation**:
- **Files used**:
  - `cnf-fcen-csv.zip`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### CancerDrugsDB

CancerDrugs_DB is a curated listing of licensed cancer drugs produced by the Anticancer Fund. Source data comes from the NCI, FDA, EMA and other data sources.

- **Website**: [https://www.anticancerfund.org/en/cancerdrugs-db](https://www.anticancerfund.org/en/cancerdrugs-db)
- **License**:
- **Citation**: ```Pantziarka P, Capistrano I R, De Potter A, Vandeborne L, Bouche G. An open access database of licensed cancer drugs. Front Pharmacol. 2021;12:627574. doi:``` [10.3389/fphar.2021.627574](http://dx.doi.org/10.3389/fphar.2021.627574)
- **Files used**:
  - `cancerdrugsdb.txt`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### CellOntology

- **Website**: [https://github.com/obophenotype/cell-ontology](https://github.com/obophenotype/cell-ontology)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**:
- **Files used**:
  - At the moment, the ontology is downloaded in OBO format (```cl-full.obo```).

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### ClinicalTrials.gov

- **Website**: [https://clinicaltrials.gov](https://clinicaltrials.gov)
- **License**: [https://clinicaltrials.gov/ct2/about-site/terms-conditions#Use](https://clinicaltrials.gov/ct2/about-site/terms-conditions)
- **Citation**:
- **Files used**:
  - `AllPublicXML.zip` (Latest monthly archive file)

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### CMAUP

CMAUP is a database of molecular activities of useful plants.

- **Website**: [http://bidd.group/CMAUP/](http://bidd.group/CMAUP/)
- **License**:
- **Citation**: ```Xian Zeng, Peng Zhang, Yali Wang, et al. CMAUP: a database of collective molecular activities of useful plants. Nucleic Acids Research 2019; 47(D1): D1118-D1127; PMID: 30357356; DOI:``` [10.1093/nar/gky965](https://dx.doi..org/10.1093/nar/gky965)
- **Files used**:
  - `CMAUPv1.0_download_Targets.txt`
  - `CMAUPv1.0_download_Plants.txt`
  - `CMAUPv1.0_download_Ingredients_onlyActive.txt`
  - `CMAUPv1.0_download_Plant_Ingredient_Associations_onlyActiveIngredients.txt`
  - `CMAUPv1.0_download_Ingredient_Target_Associations_ActivityValues_References.txt`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### CPDB

- **Website**: [http://cpdb.molgen.mpg.de](http://cpdb.molgen.mpg.de)
- **License**:
- **Citation**:
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### DGIdb

- **Website**: [http://dgidb.org](http://dgidb.org)
- **License**: The source code for DGIdb is open-source and made available under the MIT license. The license is distributed with the source code (DGIdb license). The data used in DGIdb is all open access and where possible made available as raw data dumps in the downloads section.
- **Citation**: ```DGIdb 3.0: a redesign and expansion of the drug–gene interaction database. Cotto KC, Wagner AH, Feng Y, Kiwala S, Coffman AC, Spies G, Wollam A, Spies NC, Griffith OL, Griffith M. Nucleic Acids Research. 2017 Nov 17. doi:``` [10.1093/nar/gkx1143](https://doi.org/10.1093/nar/gkx1143)
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### DiseaseOntology

- **Website**: [https://disease-ontology.org](https://disease-ontology.org)
- **License**: CC0 1.0
- **Citation**: ```Schriml LM, Munro JB, Schor M, et al. The human disease ontology 2022 update. Nucleic Acids Research. 2022;50(D1):D1255-D1261. doi:``` [10.1093/nar/gkab1063](https://doi.org/10.1093/nar/gkab1063)
- **Files used**:
  - At the moment, the ontology is downloaded in OBO format (```doid.obo```).

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### DISEASES

Gene-Disease association database

- **Website**: [https://diseases.jensenlab.org](https://diseases.jensenlab.org)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**: ```Pletscher-Frankild S, Pallejà A, Tsafou K, Binder JX, Jensen LJ. DISEASES: Text mining and data integration of disease–gene associations. Methods. 2015;74:83-89. doi:``` https://doi.org/10.1016/j.ymeth.2014.11.020
- **Files used**:
  - `human_disease_textmining_filtered.tsv`
  - `human_disease_knowledge_filtered.tsv`
  - `human_disease_experiments_filtered.tsv`
  - `human_disease_integrated_full.tsv`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### DrugBank

DrugBank Online is a comprehensive, free-to-access, online database containing information on drugs and drug targets.

- **Website**: [https://go.drugbank.com](https://go.drugbank.com)
- **License**:
  - Creative Common’s Attribution-NonCommercial 4.0 International License.
  - DrugBank is offered to the public as a freely available resource.
  - DrugBank Online is offered to the public as a free-to-access resource. Use and re-distribution of the content of DrugBank Online or the DrugBank Data, in whole or in part, for any purpose requires a license. Academic users may apply for a free license for certain use cases and all other users require a paid license.
- **Citation**: ```Wishart DS, Feunang YD, Guo AC, Lo EJ, Marcu A, Grant JR, Sajed T, Johnson D, Li C, Sayeeda Z, Assempour N, Iynkkaran I, Liu Y, Maciejewski A, Gale N, Wilson A, Chin L, Cummings R, Le D, Pon A, Knox C, Wilson M. DrugBank 5.0: a major update to the DrugBank database for 2018. Nucleic Acids Res. 2017 Nov 8. doi:``` [10.1093/nar/gkx1037](https://doi.org/10.1093/nar/gkx1037)
- **Files used**:
  - `drugbank_all_full_database.xml.zip`
  - `drugbank_all_structures.sdf.zip`
  - `drugbank_all_metabolite-structures.sdf.zip`

**Configuration properties**:

| Property             | Values     | Description                                                               |
|----------------------|------------|---------------------------------------------------------------------------|
| forceExport          | true/false | Force graph export even if nothing changed                                |
| skipDrugInteractions | true/false | Skip exporting of drug interactions which currently take the longest time |

----

### DrugCentral

DrugCentral is an online drug information resource created and maintained by the Division of Translational Informatics at University of New Mexico in collaboration with the IDG.

- **Website**: [https://drugcentral.org](https://drugcentral.org)
- **License**: CC BY-SA 4.0
- **Citation**: ```Ursu O, Holmes J, Knockel J, et al. DrugCentral: online drug compendium. Nucleic Acids Res. 2017;45(D1):D932-D939. doi:``` [10.1093/nar/gkw993](https://doi.org/10.1093/nar/gkw993)
- **Files used**:
  - The SQL database dump file `drugcentral-pgdump_[YYYYMMDD].sql.gz` is used to extract the database tables into TSV files.

**Configuration properties**:

| Property               | Values     | Description                                                                   |
|------------------------|------------|-------------------------------------------------------------------------------|
| forceExport            | true/false | Force graph export even if nothing changed                                    |
| skipDrugLabelFullTexts | true/false | Skip export of drug label full text information to reduce graph database size |
| skipLINCSSignatures    | true/false | Skip export of LINCS signatures to reduce graph database size                 |
| skipFAERSReports       | true/false | Skip export of FAERS reports to reduce graph database size                    |

----

### EFO

The Experimental Factor Ontology (EFO) provides a systematic description of many experimental variables including diseases.

- **Website**: [https://www.ebi.ac.uk/efo/](https://www.ebi.ac.uk/efo/)
- **License**: Apache-2.0
- **Citation**: ```Malone J, Holloway E, Adamusiak T, Kapushesky M, Zheng J, Kolesnikov N, Zhukova A, Brazma A, Parkinson H: Modeling Sample Variables with an Experimental Factor Ontology. Bioinformatics 2010, 26(8):1112-1118. doi:``` [10.1093/bioinformatics/btq099](https://doi.org/10.1093/bioinformatics/btq099)
- **Files used**:
  - At the moment, the ontology is downloaded in OBO format (```efo.obo```).

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### EMA

European Medicines Agency's (EMA) medicine-related data [EPAR and HMPC]

- **Website**: [https://www.ema.europa.eu/en/medicines/download-medicine-data](https://www.ema.europa.eu/en/medicines/download-medicine-data)
- **License**: Used data are publicly available on the EMA website at https://www.ema.europa.eu/en/medicines/download-medicine-data.
  - Access to documents policy can be found at https://www.ema.europa.eu/en/about-us/how-we-work/access-documents
- **Citation**:
- **Files used**:
  - `Medicines_output_european_public_assessment_reports.xlsx`
  - `Medicines_output_herbal_medicines.xlsx`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### ENZYME

Expasy ENZYME

- **Website**: [https://enzyme.expasy.org](https://enzyme.expasy.org)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**: ```Bairoch A. The ENZYME database in 2000. Nucleic Acids Research. 2000;28(1):304-305. doi:``` [10.1093/nar/28.1.304](http://doi.org/10.1093/nar/28.1.304)
- **Files used**:
  - `enzyme.dat`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### GenCC

- **Website**: [https://thegencc.org](https://thegencc.org)
- **License**: [GenCC Terms of Use](https://thegencc.org/terms.html)
  - The GenCC data are available free of restriction under a CC0 1.0 Universal (CC0 1.0) Public Domain Dedication. The GenCC requests that you give attribution to GenCC and the contributing sources whenever possible and appropriate.
- **Citation**: ```DiStefano, Marina T. et al. The Gene Curation Coalition: A global effort to harmonize gene–disease evidence resources. Genetics in Medicine, Volume 24, Issue 8, 1732 - 1742. doi:``` [10.1016/j.gim.2022.04.017](https://doi.org/10.1016/j.gim.2022.04.017)
- **Files used**:
  - `gencc-submissions.tsv`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### Gene2Phenotype

G2P is a publicly-accessible online system designed to facilitate the development, validation, curation and distribution of large-scale, evidence-based datasets for use in diagnostic variant filtering. Each G2P entry associates an allelic requirement and a mutational consequence at a defined locus with a disease entity. A confidence level and evidence link are assigned to each entry.

- **Website**: [https://www.ebi.ac.uk/gene2phenotype](https://www.ebi.ac.uk/gene2phenotype)
- **License**: Apache License 2.0
- **Citation**:
- **Files used**:
  - `CancerG2P.csv.gz`
  - `DDG2P.csv.gz`
  - `EyeG2P.csv.gz`
  - `SkinG2P.csv.gz`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### GeneOntology

The Gene Ontology (GO) knowledgebase is the world’s largest source of information on the functions of genes. This knowledge is both human-readable and machine-readable, and is a foundation for computational analysis of large-scale molecular biology and genetics experiments in biomedical research.

- **Website**: [http://geneontology.org](http://geneontology.org)
- **License**: Gene Ontology Consortium data and data products are licensed under the [Creative Commons Attribution 4.0 Unported License](https://creativecommons.org/licenses/by/4.0/legalcode)
- **Citation**: For citation information see http://geneontology.org/docs/go-citation-policy/
- **Files used**:
  - `go.obo`
  - `goa_human.gaf.gz`
  - `goa_human_complex.gaf.gz`
  - `goa_human_isoform.gaf.gz`
  - `goa_human_rna.gaf.gz`

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### GuideToPharmacology

The International Union of Basic and Clinical Pharmacology (IUPHAR) / British Pharmacological Society (BPS) Guide to
PHARMACOLOGY is an expert-curated resource of ligand-activity-target relationships, the majority of which come from
high-quality pharmacological and medicinal chemistry literature.

- **Website**: [https://www.guidetopharmacology.org](https://www.guidetopharmacology.org)
- **License**: [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)
- **Citation**: ```Harding SD, Armstrong JF, Faccenda E, Southan C, Alexander SPH, Davenport AP, Pawson AJ, Spedding M, Davies JA; NC-IUPHAR. (2021) The IUPHAR/BPS guide to PHARMACOLOGY in 2022: curating pharmacology for COVID-19, malaria and antibacterials. Nucl. Acids Res. Online ahead of print. doi:``` [10.1093/nar/gkab1010](https://doi.org/10.1093/nar/gkab1010)
  - [https://www.guidetopharmacology.org/citing.jsp](https://www.guidetopharmacology.org/citing.jsp)
- **Files used**:
  - `public_iuphardb_vXXXX.X.zip`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### GWASCatalog

The NHGRI-EBI Catalog of human genome-wide association studies

- **Website**: [https://www.ebi.ac.uk/gwas/home](https://www.ebi.ac.uk/gwas/home)
- **License**: The NHGRI-EBI GWAS Catalog and all its contents are available under the general [Terms of Use for EMBL-EBI Services](http://www.ebi.ac.uk/about/terms-of-use).
- **Citation**: ```Buniello A, MacArthur JAL, Cerezo M, Harris LW, Hayhurst J, Malangone C, McMahon A, Morales J, Mountjoy E, Sollis E, Suveges D, Vrousgou O, Whetzel PL, Amode R, Guillen JA, Riat HS, Trevanion SJ, Hall P, Junkins H, Flicek P, Burdett T, Hindorff LA, Cunningham F and Parkinson H.
The NHGRI-EBI GWAS Catalog of published genome-wide association studies, targeted arrays and summary statistics 2019.
Nucleic Acids Research, 2019, Vol. 47 (Database issue): D1005-D1012.```
- **Files used**:
  - `gwas_catalog_v1.0.2-associations.tsv`
  - `gwas_catalog_v1.0.2-studies.tsv`
  - `gwas_catalog-ancestry.tsv`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### HAMAP

Expasy HAMAP

- **Website**: [https://hamap.expasy.org](https://hamap.expasy.org)
- **License**:
- **Citation**:
- **Files used**:
  - `hamap_rules.dat`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### HERB

The HGNC is responsible for approving unique symbols and names for human loci, including protein coding genes, ncRNA genes and pseudogenes, to allow unambiguous scientific communication.

- **Website**: [http://herb.ac.cn](http://herb.ac.cn)
- **License**:
- **Citation**: ```Fang S, Dong L, Liu L, et al. HERB: a high-throughput experiment- and reference-guided database of traditional Chinese medicine. Nucleic Acids Research. 2021;49(D1):D1197-D1206. doi:``` [10.1093/nar/gkaa1063](https://doi.org/10.1093/nar/gkaa1063)
- **Files used**:
  - `HERB_disease_info.txt`
  - `HERB_experiment_info.txt`
  - `HERB_herb_info.txt`
  - `HERB_ingredient_info.txt`
  - `HERB_reference_info.txt`
  - `HERB_target_info.txt`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### HGNC

The HGNC is responsible for approving unique symbols and names for human loci, including protein coding genes, ncRNA genes and pseudogenes, to allow unambiguous scientific communication.

- **Website**: [https://www.genenames.org](https://www.genenames.org)
- **License**: https://www.genenames.org/about/
  - **Data release policy**: No restrictions are imposed on access to, or use of, the data provided by the HGNC, which are provided to enhance knowledge and encourage progress in the scientific community. The HGNC provide these data in good faith, but make no warranty, express or implied, nor assume any legal liability or responsibility for any purpose for which they are used.
  - **Guidelines on use of data in publications (copyright and licensing)**: It is a condition of our funding from NIH and the Wellcome Trust that the nomenclature and information we provide is freely available to all. Anyone may use the HGNC data, but we request that they reference the "HUGO Gene Nomenclature Committee at the European Bioinformatics Institute" and the website where possible.
- **Citation**: ```Tweedie S, Braschi B, Gray KA, Jones TEM, Seal RL, Yates B, Bruford EA. Genenames.org: the HGNC and VGNC resources in 2021. Nucleic Acids Res. PMID: 33152070 PMCID: PMC7779007 doi:``` [10.1093/nar/gkaa980](https://doi.org/10.1093/nar/gkaa980)
- **Files used**:
  -`hgnc_complete_set.txt`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### HMDB

- **Website**: [https://hmdb.ca](https://hmdb.ca)
- **License**: ```HMDB is offered to the public as a freely available resource. Use and re-distribution of the data, in whole or in part, for commercial purposes requires explicit permission of the authors and explicit acknowledgment of the source material (HMDB) and the original publication (see below). We ask that users who download significant portions of the database cite the HMDB paper in any resulting publications.```
- **Citation**: ```Wishart DS, Guo AC, Oler E, et al., HMDB 5.0: the Human Metabolome Database for 2022. Nucleic Acids Res. 2022. Jan 7;50(D1):D622–31. PMID: 34986597.```
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### HPO

The Human Phenotype Ontology (HPO) provides a standardized vocabulary of phenotypic abnormalities encountered in human disease.

- **Website**: [https://hpo.jax.org](https://hpo.jax.org)
- **License**: License information for HPO can be found at [https://hpo.jax.org/app/license](https://hpo.jax.org/app/license).
- **Citation**: ```Sebastian Köhler, Michael Gargano, Nicolas Matentzoglu, Leigh C Carmody, David Lewis-Smith, Nicole A Vasilevsky, Daniel Danis, Ganna Balagura, Gareth Baynam, Amy M Brower, Tiffany J Callahan, Christopher G Chute, Johanna L Est, Peter D Galer, Shiva Ganesan, Matthias Griese, Matthias Haimel, Julia Pazmandi, Marc Hanauer, Nomi L Harris, Michael J Hartnett, Maximilian Hastreiter, Fabian Hauck, Yongqun He, Tim Jeske, Hugh Kearney, Gerhard Kindle, Christoph Klein, Katrin Knoflach, Roland Krause, David Lagorce, Julie A McMurry, Jillian A Miller, Monica C Munoz-Torres, Rebecca L Peters, Christina K Rapp, Ana M Rath, Shahmir A Rind, Avi Z Rosenberg, Michael M Segal, Markus G Seidel, Damian Smedley, Tomer Talmy, Yarlalu Thomas, Samuel A Wiafe, Julie Xian, Zafer Yüksel, Ingo Helbig, Christopher J Mungall, Melissa A Haendel, Peter N Robinson, The Human Phenotype Ontology in 2021, Nucleic Acids Research, Volume 49, Issue D1, 8 January 2021, Pages D1207–D1217. doi:``` [10.1093/nar/gkaa1043](https://doi.org/10.1093/nar/gkaa1043)
- **Files used**:
  - At the moment, the ontology is downloaded in OBO format (```hp.obo```).
  - Please note that the HPO OBO version lacks some of the logical reasoning which is present in the OWL format version ([info](https://hpo.jax.org/app/download/ontology)). Once a suitable OWL parser is implemented, the ontology file will likely be switched!
  - For annotations, the files ```phenotype.hpoa``` and ```phenotype_to_genes.txt``` are parsed and used in the graph exporter.
  - The file ```genes_to_phenotype.txt``` is stated to be analogous to ```phenotype_to_genes.txt``` although the file size differs dramatically. Therefore, this file is currently ignored.

**Configuration properties**:

| Property       | Values     | Description                                                                                                                                                               |
|----------------|------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed                                                                                                                                |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete                                                                                                                  |
| omimLicensed   | true/false | Annotations with evidence IEA require contacting OMIM ([info](https://hpo.jax.org/app/help/annotations)). By setting this flag to true, you agree that this has happened! |

----

### HPRD

- **Website**: [http://www.hprd.org](http://www.hprd.org)
- **License**: ```Everything in HPRD is free as long as it is not used for commercial purposes. Commercial entitites will have to pay a fee under a licensing arrangement which will be used to make this database even better. Commercial users should send an e-mail for details. This model of HPRD is similar to the SWISS-PROT licensing arrangement.``` ([http://hprd.org/FAQ](http://hprd.org/FAQ))
- **Citation**:
  - ```Peri S, Navarro JD, Amanchy R, et al. Development of human protein reference database as an initial platform for approaching systems biology in humans. Genome Res. 2003;13(10):2363-2371. doi:``` [10.1101/gr.1680803](http://doi.org/10.1101/gr.1680803)
  - ```Keshava Prasad TS, Goel R, Kandasamy K, et al. Human protein reference database--2009 update. Nucleic Acids Research. 2009;37(Database):D767-D772. doi:``` [10.1093/nar/gkn892](http://doi.org/10.1093/nar/gkn892)
- **Files used**:
  - `HPRD_FLAT_FILES_[...].tar.gz`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### IntAct

- **Website**: [https://www.ebi.ac.uk/intact/](https://www.ebi.ac.uk/intact/)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**:
- **Files used**:
  - `psi30/species/human.zip`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### InterPro

InterPro provides functional analysis of proteins by classifying them into families and predicting domains and important sites.

- **Website**: [https://www.ebi.ac.uk/interpro/](https://www.ebi.ac.uk/interpro/)
- **License**:
- **Citation**: ```Blum M, Chang H, Chuguransky S, Grego T, Kandasaamy S, Mitchell A, Nuka G, Paysan-Lafosse T, Qureshi M, Raj S, RichardsonL, Salazar GA, Williams L, Bork P, Bridge A, Gough J, Haft DH, Letunic I, Marchler-Bauer A, Mi H, Natale DA, Necci M, Orengo CA, Pandurangan AP, Rivoire C, Sigrist CJA, Sillitoe I, Thanki N, Thomas PD, Tosatto SCE, Wu CH, Bateman A and Finn RD The InterPro protein families and domains database: 20 years on. Nucleic Acids Research, Nov 2020. doi:``` [10.1093/nar/gkaa977](https://doi.org/10.1093/nar/gkaa977)
- **Files used**:
  - `interpro.xml.gz`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### ITIS

Integrated Taxonomic Information System

- **Website**: [https://www.itis.gov](https://www.itis.gov)
- **License**: CC0
  - Information presented on the ITIS website is considered public information and may be distributed or copied. Use of the appropriate citation is requested.
- **Citation**:
  - To direct readers to the entire ITIS website, it is sufficient to give the address of the site in text:
  - Retrieved [month, day, year], from the Integrated Taxonomic Information System (ITIS), www.itis.gov. doi: [10.5066/F7KH0KBK](https://doi.org/10.5066/F7KH0KBK)
  - To cite data obtained from ITIS, the following citation format is offered as a suggestion:
  - Retrieved [month, day, year], from the Integrated Taxonomic Information System (ITIS) on-line database, www.itis.gov. doi: [10.5066/F7KH0KBK](https://doi.org/10.5066/F7KH0KBK)
  - Citations can be formatted to the style needed for your publication as long as there is sufficient information included to return the reader to the referenced information.
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### KEGG

- **Website**: [https://www.kegg.jp](https://www.kegg.jp)
- **License**: A subset of KEGG MEDICUS is freely available at the GenomeNet FTP site. ftp://ftp.genome.jp/pub/kegg/medicus/
- **Citation**:
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### MammalianPhenotypeOntology

- **Website**: [https://github.com/mgijax/mammalian-phenotype-ontology](https://github.com/mgijax/mammalian-phenotype-ontology)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**:
- **Files used**:
  - `mp-full.obo`

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### MassSpectrometryOntology

- **Website**: [https://github.com/HUPO-PSI/psi-ms-CV](https://github.com/HUPO-PSI/psi-ms-CV)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**: ```Mayer G, Montecchi-Palazzi L, Ovelleiro D, Jones AR, Binz PA, Deutsch EW, Chambers M, Kallhardt M, Levander F, Shofstahl J, Orchard S, Vizcaíno JA, Hermjakob H, Stephan C, Meyer HE, Eisenacher M; HUPO-PSI Group. The HUPO proteomics standards initiative- mass spectrometry controlled vocabulary. Database (Oxford). 2013 Mar 12;2013:bat009. doi:``` [10.1093/database/bat009](https://doi.org/10.1093/database/bat009)
- **Files used**:
  - `psi-ms.obo`

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### MED-RT

- **Website**: [https://evs.nci.nih.gov/ftp1/MED-RT/](https://evs.nci.nih.gov/ftp1/MED-RT/)
- **License**:
  - https://evs.nci.nih.gov/ftp1/MED-RT/medrt_about.html
  - MED-RT has no copyright acknowledgement required. However, users must adhere to the UMLS license.
- **Citation**:
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### miR2Disease

- **Website**: [http://www.mir2disease.org](http://www.mir2disease.org)
- **License**:
- **Citation**:
- **Files used**:
  - ```miRtar.txt```
  - ```diseaseList.txt```
  - ```AllEntries.txt```

----

### miRBase

- **Website**: [https://www.mirbase.org](https://www.mirbase.org)
- **License**: CC0 with attribution
- **Citation**: ```Kozomara A, Birgaoanu M, Griffiths-Jones S. miRBase: from microRNA sequences to function. Nucleic Acids Res 2019 47:D155-D162. doi:``` [10.1093/nar/gky1141](https://doi.org/10.1093/nar/gky1141)
- **Files used**:
  - `confidence.txt.gz`
  - `confidence_score.txt.gz`
  - `dead_mirna.txt.gz`
  - `hairpin.fa.gz`
  - `mature.fa.gz`
  - `mature_database_links.txt.gz`
  - `mature_database_url.txt.gz`
  - `miRNA.dat.gz`
  - `miRNA.str.gz`
  - `mirna.txt.gz`
  - `mirna_2_prefam.txt.gz`
  - `mirna_chromosome_build.txt.gz`
  - `mirna_context.txt.gz`
  - `mirna_database_links.txt.gz`
  - `mirna_database_url.txt.gz`
  - `mirna_mature.txt.gz`
  - `mirna_prefam.txt.gz`
  - `mirna_pre_mature.txt.gz`
  - `mirna_species.txt.gz`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### miRDB

- **Website**: [https://mirdb.org](https://mirdb.org)
- **License**:
- **Citation**:
  - ```Yuhao Chen and Xiaowei Wang (2020) miRDB: an online database for prediction of functional microRNA targets. Nucleic Acids Research. 48(D1):D127-D131. doi:``` [10.1093/nar/gkz757](https://doi.org/10.1093/nar/gkz757)
  - ```Weijun Liu and Xiaowei Wang (2019) Prediction of functional microRNA targets by integrative modeling of microRNA binding and target expression data. Genome Biology. 20(1):18. doi:``` [10.1186/s13059-019-1629-z](https://doi.org/10.1186/s13059-019-1629-z)
- **Files used**:
  - `miRDB_prediction_result.txt.gz`

**Configuration properties**:

| Property       | Values       | Description                                           |
|----------------|--------------|-------------------------------------------------------|
| forceExport    | true/false   | Force graph export even if nothing changed            |
| scoreThreshold | [0.0, 100.0] | Filter threshold for prediction score (default: 50.0) |

----

### miRTarBase

- **Website**: [https://mirtarbase.cuhk.edu.cn](https://mirtarbase.cuhk.edu.cn)
- **License**: [https://mirtarbase.cuhk.edu.cn/~miRTarBase/miRTarBase_2022/cache/download/LICENSE](https://mirtarbase.cuhk.edu.cn/~miRTarBase/miRTarBase_2022/cache/download/LICENSE)
- **Citation**: ```miRTarBase update 2022: an informative resource for experimentally validated miRNA–target interactions. Nucleic Acids Research 2022 Jan 7;50(D1):D222-D230. doi:``` [10.1093/nar/gkab1079](https://doi.org/10.1093/nar/gkab1079)
- **Files used**:
  - `miRTarBase_MTI.xlsx`
  - `MicroRNA_Target_Sites.xlsx`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### MolecularInteractionsOntology

- **Website**: [https://github.com/HUPO-PSI/psi-mi-CV](https://github.com/HUPO-PSI/psi-mi-CV)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**:
- **Files used**:
  - `psi-mi.obo`

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### Mondo

- **Website**: [https://mondo.monarchinitiative.org](https://mondo.monarchinitiative.org)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**:
- **Files used**:
  - `mondo.obo`

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### NDF-RT

- **Website**: [https://evs.nci.nih.gov/ftp1/NDF-RT/](https://evs.nci.nih.gov/ftp1/NDF-RT/)
- **License**: https://www.nlm.nih.gov/research/umls/sourcereleasedocs/current/NDFRT/index.html
  - NDF-RT is made available via the UMLS. Users of all UMLS ontologies must abide by the terms of the UMLS license, available at https://uts.nlm.nih.gov/license.html
- **Citation**:
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### Negatome

The Negatome is a collection of protein and domain pairs which are unlikely engaged in direct physical interactions.

- **Website**: [http://mips.helmholtz-muenchen.de/proj/ppi/negatome/](http://mips.helmholtz-muenchen.de/proj/ppi/negatome/)
- **License**:
- **Citation**: ```Blohm P, Frishman G, Smialowski P, et al. Negatome 2.0: a database of non-interacting proteins derived by literature mining, manual annotation and protein structure analysis. Nucl Acids Res. 2014;42(D1):D396-D400. doi:``` [10.1093/nar/gkt1079](https://doi.org/10.1093/nar/gkt1079)
- **Files used**:
  - `manual.txt`
  - `manual_stringent.txt`
  - `manual_pfam.txt`
  - `pdb.txt`
  - `pdb_stringent.txt`
  - `pdb_pfam.txt`
  - `combined.txt`
  - `combined_stringent.txt`
  - `combined_pfam.txt`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### NeuroBehaviourOntology

- **Website**: [https://github.com/obo-behavior/behavior-ontology](https://github.com/obo-behavior/behavior-ontology)
- **License**: 
- **Citation**:
- **Files used**:
  - At the moment, the ontology is downloaded in OBO format (```nbo-full.obo```).

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### OMIM

Online Mendelian Inheritance in Man® - An Online Catalog of Human Genes and Genetic Disorders

- **Website**: [https://www.omim.org](https://www.omim.org)
- **License**: Research and educational use of OMIM is encouraged, and single-user academic, non-profit, and governmental agencies can register for access to the files without a license.
  - Johns Hopkins University holds the copyright to OMIM including the collective data therein. I hereby declare that I will not use or share the data contained in OMIM for any commercial purposes, will not develop a derivative database, nor distribute the data to a third party without first obtaining a license from Johns Hopkins University to do so. If OMIM data are used in research, I will notify OMIM and send a copy of the reference.
- **Citation**:
  - ```Amberger JS, Bocchini CA, Scott AF, Hamosh A. OMIM.org: leveraging knowledge across phenotype-gene relationships. Nucleic Acids Res. 2019 Jan 8;47(D1):D1038-D1043. doi: [10.1093/nar/gky1151](https://doi.org/10.1093/nar/gky1151). PMID: 30445645.```
  - ```Amberger JS, Bocchini CA, Schiettecatte FJM, Scott AF, Hamosh A. OMIM.org: Online Mendelian Inheritance in Man (OMIM®), an online catalog of human genes and genetic disorders. Nucleic Acids Res. 2015 Jan;43(Database issue):D789-98. PMID: 25428349. doi:``` [10.1093/nar/gku1205](https://doi.org/10.1093/nar/gku1205).
  - ```Online Mendelian Inheritance in Man, OMIM®. McKusick-Nathans Institute of Genetic Medicine, Johns Hopkins University (Baltimore, MD), {date}. World Wide Web URL: https://omim.org/```
- **Files used**:
  - `genemap2.txt`
  - `mimtitles.txt`

**Configuration properties**:

| Property    | Values     | Description                                                                      |
|-------------|------------|----------------------------------------------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed                                       |
| downloadKey |            | License key obtained by registering with OMIM.org required for downloading files |

----

### OpenTargets

Open Targets is an innovative, large-scale, multi-year, public-private partnership that uses human genetics and genomics data for systematic drug target identification and prioritisation.

- **Website**: [https://www.opentargets.org](https://www.opentargets.org)
- **License**: Datasets generated by the Open Targets Platform are freely available for download for academic or commercial purposes to support further downstream analyses ([Link](https://platform-docs.opentargets.org/licence)).
- **Citation**: ```Ochoa, D. et al. (2021). Open Targets Platform: supporting systematic drug–target identification and prioritisation. Nucleic Acids Research. doi:``` [10.1093/nar/gkaa1027](https://doi.org/10.1093/nar/gkaa1027)
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### PathwayCommons

Pathway Commons is an integrated resource of publicly available information about biological pathways including biochemical reactions, assembly of biomolecular complexes, transport and catalysis events and physical interactions involving proteins, DNA, RNA, and small molecules (e.g. metabolites and drug compounds).

- **Website**: [http://www.pathwaycommons.org](http://www.pathwaycommons.org)
- **License**:  All of the data provided by Pathway Commons is free. In particular, Pathway Commons distributes pathway information with the intellectual property restrictions of the source database; Only databases that are freely available for academics are included.
  - For further license information on source databases see [https://www.pathwaycommons.org/pc2/datasources](https://www.pathwaycommons.org/pc2/datasources).
- **Citation**: ```Rodchenkov I, Babur O, Luna A, et al. Pathway Commons 2019 Update: integration, analysis and exploration of pathway data. Nucleic Acids Research. Published online October 24, 2019:gkz946. doi:``` [10.1093/nar/gkz946](https://doi.org/10.1093/nar/gkz946)
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### PharmGKB

- **Website**: [https://www.pharmgkb.org](https://www.pharmgkb.org)
- **License**: https://www.pharmgkb.org/page/dataUsagePolicy
- **Citation**: https://www.pharmgkb.org/page/citingPharmgkb
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### PhenotypeAndTraitOntology

- **Website**: [https://github.com/pato-ontology/pato](https://github.com/pato-ontology/pato)
- **License**: [CC BY 3.0](https://creativecommons.org/licenses/by/3.0/)
- **Citation**:
- **Files used**:
  - `patofull.obo`

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### PROSITE

Expasy PROSITE and PRORULE

- **Website**: [https://prosite.expasy.org](https://prosite.expasy.org)
- **License**: [CC BY-NC-ND 4.0](https://creativecommons.org/licenses/by-nc-nd/4.0/)
- **Citation**:
  - ```Sigrist CJA, de Castro E, Cerutti L, et al. New and continuing developments at PROSITE. Nucleic Acids Research. 2012;41(D1):D344-D347. doi:``` [10.1093/nar/gks1067](http://doi.org/10.1093/nar/gks1067)
  - ```Sigrist CJA, De Castro E, Langendijk-Genevaux PS, Le Saux V, Bairoch A, Hulo N. ProRule: a new database containing functional and structural information on PROSITE profiles. Bioinformatics. 2005;21(21):4060-4066. doi:``` [10.1093/bioinformatics/bti614](http://doi.org/10.1093/bioinformatics/bti614)
- **Files used**:
  - `prosite.dat`
  - `prorule.dat`
  - `prosite.doc`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### ReDO-DB

ReDO_DB is a curated listing of non-cancer drugs which have shown some evidence of anticancer activity produced by the Anticancer Fund.

- **Website**: [https://www.anticancerfund.org/en/redo-db](https://www.anticancerfund.org/en/redo-db)
- **License**:
- **Citation**: ```Pantziarka P, Verbaanderd C, Sukhatme V, et al. ReDO_DB: the repurposing drugs in oncology database. ecancer. 2018;12. doi:``` [10.3332/ecancer.2018.886](https://doi.org/10.3332/ecancer.2018.886)
- **Files used**:
  - `ReDO_DB.txt`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### ReDOTrialsDB

ReDOTrialsDB is a curated database, produced by the Anticancer Fund, of active clinical trials investigating the use of non-cancer drugs as potential cancer treatments

- **Website**: [https://www.anticancerfund.org/en/redo-trials-db](https://www.anticancerfund.org/en/redo-trials-db)
- **License**:
- **Citation**: ```Pantziarka, P., Vandeborne, L., & Bouche, G. (2021). A Database of Drug Repurposing Clinical Trials in Oncology. (June 21, 2021) doi:``` [10.31219/osf.io/yagft](https://doi.org/10.31219/osf.io/yagft)
- **Files used**:
  - `ReDO_Trials_DB.txt`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### RefSeq

- **Website**: [https://www.ncbi.nlm.nih.gov/refseq/](https://www.ncbi.nlm.nih.gov/refseq/)
- **License**: Public domain except for third-party information (https://www.ncbi.nlm.nih.gov/home/about/policies/)
- **Citation**: ```O'Leary NA, Wright MW, Brister JR, Ciufo S, Haddad D, McVeigh R, Rajput B, Robbertse B, Smith-White B, Ako-Adjei D, Astashyn A, Badretdin A, Bao Y, Blinkova O, Brover V, Chetvernin V, Choi J, Cox E, Ermolaeva O, Farrell CM, Goldfarb T, Gupta T, Haft D, Hatcher E, Hlavina W, Joardar VS, Kodali VK, Li W, Maglott D, Masterson P, McGarvey KM, Murphy MR, O'Neill K, Pujar S, Rangwala SH, Rausch D, Riddick LD, Schoch C, Shkeda A, Storz SS, Sun H, Thibaud-Nissen F, Tolstoy I, Tully RE, Vatsan AR, Wallin C, Webb D, Wu W, Landrum MJ, Kimchi A, Tatusova T, DiCuccio M, Kitts P, Murphy TD, Pruitt KD. Reference sequence (RefSeq) database at NCBI: current status, taxonomic expansion, and functional annotation. Nucleic Acids Res. 2016 Jan 4;44(D1):D733-45. doi:``` [10.1093/nar/gkv1189](http://doi.org/10.1093/nar/gkv1189)
- **Files used**:
  - `[...]_genomic.gff.gz` depending on configured assembly

**Configuration properties**:

| Property    | Values     | Description                                     |
|-------------|------------|-------------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed      |
| assembly    |            | Assembly name to be used. Example: `GRCh38.p14` |

----

### RNADisease

- **Website**: [http://www.rnadisease.org](http://www.rnadisease.org)
- **License**: non-commercial with attribution
- **Citation**: ```Chen J, Lin J, Hu Y, et al. RNADisease v4.0: an updated resource of RNA-associated diseases, providing RNA-disease analysis, enrichment and prediction. Nucleic Acids Research. 2023;51(D1):D1397-D1404. doi:``` [10.1093/nar/gkac814](http://doi.org/10.1093/nar/gkac814)
- **Files used**:
  - `RNA-disease_experiment_all.zip`
  - `RNA-disease_miRNA_predict.zip`
  - `RNA-disease_lncRNA_predict.zip`
  - `RNA-disease_circRNA_predict.zip`
  - `RNA-disease_piRNA_predict.zip`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### RNAInter

- **Website**: [http://www.rnainter.org](http://www.rnainter.org)
- **License**:  Data is provided for non-commercial use, distribution, or reproduction in any medium, only if you properly cite the original work. If you are a commercial user, please contact the authors firstly.
- **Citation**: ```Kang J, Tang Q, He J, et al. RNAInter v4.0: RNA interactome repository with redefined confidence scoring system and improved accessibility. Nucleic Acids Research. 2022;50(D1):D326-D332. doi:``` [10.1093/nar/gkab997](https://doi.org/10.1093/nar/gkab997)
- **Files used**:
  - `Download_data_RC.tar.gz`
  - `Download_data_RD.tar.gz`
  - `Download_data_RH.tar.gz`
  - `Download_data_RP.tar.gz`
  - `Download_data_RR.tar.gz`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### RNALocate

RNA subcellular localization database

- **Website**: [http://www.rna-society.org/rnalocate/](http://www.rna-society.org/rnalocate/)
- **License**:  RNALocate contains data for non-commercial use, distribution, and reproduction in any medium, provided the original work is properly cited. If you are commercial user, please contact the RNALocate team.
- **Citation**: ```Cui T, Dou Y, Tan P, et al. RNALocate v2.0: an updated resource for RNA subcellular localization with increased coverage and annotation. Nucleic Acids Research. 2022;50(D1):D333-D339. doi:``` [10.1093/nar/gkab825](https://doi.org/10.1093/nar/gkab825)
- **Files used**:
  - `All experimental RNA subcellular localization data.zip`
  - `Other RNA subcellular localization data.zip`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### SequenceOntology

- **Website**: [http://www.sequenceontology.org](http://www.sequenceontology.org)
- **License**: CC BY-SA 4.0
- **Citation**:
- **Files used**:
  - `so.obo`

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### SIDER

SIDER contains information on marketed medicines and their recorded adverse drug reactions. The information is extracted from public documents and package inserts. The available information include side effect frequency, drug and side effect classifications as well as links to further information, for example drug–target relations.

> ⚠️ Please note that SIDER hasn't been updated since 2015 (last checked 06.2021). As the information were extracted using text-mining techniques, data errors are inevitable and some are already known. However, SIDER is still one of the best, freely available side effect resources available.

## Data issues

* [carvedilol](http://sideeffects.embl.de/drugs/2585/?labels=all)
  * SIDER lists Asthma as indication, however, the [drug label](http://sideeffects.embl.de/labels/dpdonline/0TF8oZKGp0k=/C0004096/) lists Asthma under contra-indications and other data sources such as [DrugCentral](https://drugcentral.org/drugcard/522?q=carvedilol#druguse) agree.

- **Website**: [http://sideeffects.embl.de](http://sideeffects.embl.de)
- **License**:
  - The files ```meddra_all_indications.tsv.gz```, ```meddra_all_se.tsv.gz```, and ```meddra_freq.tsv.gz``` are licensed under a Creative Commons Attribution-Noncommercial-Share Alike 4.0 License [![](https://licensebuttons.net/l/by-nc-sa/4.0/80x15.png)](http://creativecommons.org/licenses/by-nc-sa/4.0/)
  - The files ```drug_names.tsv``` and ```drug_atc.tsv``` are licensed under a CC0 1.0 Universal Public Domain Dedication license [![](https://licensebuttons.net/l/zero/1.0/80x15.png)](https://creativecommons.org/publicdomain/zero/1.0/)
  - For commercial use or customized versions, please contact biobyte solutions GmbH at [http://www.biobyte.de](http://www.biobyte.de).
- **Citation**:
  - ```Kuhn M, Letunic I, Jensen LJ, Bork P. The SIDER database of drugs and side effects. Nucleic Acids Res. 2015 Oct 19. doi:``` [10.1093/nar/gkv1075](http://doi.org/10.1093/nar/gkv1075)
  - ```Kuhn M, Campillos M, Letunic I, Jensen LJ, Bork P. A side effect resource to capture phenotypic effects of drugs. Mol Syst Biol. 2010;6:343. Epub 2010 Jan 19.```
- **Files used**:
  - `meddra_all_indications.tsv.gz`
  - `meddra_all_se.tsv.gz`
  - `meddra_freq.tsv.gz`
  - `drug_names.tsv`
  - `drug_atc.tsv`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### STITCH

STITCH Chemical-Protein Interaction Networks

- **Website**: [http://stitch.embl.de](http://stitch.embl.de)
- **License**: STITCH is available for licensing — both for commercial and for academic institutions [here](http://stitch.embl.de/cgi/access.pl?footer_active_subpage=licensing).
- **Citation**: ```Szklarczyk, D., Franceschini, A., Wyder, S., Forslund, K., Heller, D., Huerta-Cepas, J., Simonovic, M., Roth, A., Santos, A., Tsafou, K. P., Kuhn, M., Bork, P., Jensen, L. J., & von Mering, C. (2014). STRING v10: protein–protein interaction networks, integrated over the tree of life. In Nucleic Acids Research (Vol. 43, Issue D1, pp. D447–D452). Oxford University Press (OUP). doi:``` [10.1093/nar/gku1003](https://doi.org/10.1093/nar/gku1003)
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### STRING

STRING Protein-Protein Interaction Networks

- **Website**: [https://string-db.org](https://string-db.org)
- **License**: All data and download files in STRING are freely available under a 'Creative Commons BY 4.0' license.
  - When using the data, please provide appropriate credit — and inform users of any changes or additions that you might have made to the data.
  - [https://string-db.org/cgi/access?footer_active_subpage=licensing](https://string-db.org/cgi/access?footer_active_subpage=licensing)
- **Citation**: ```Szklarczyk, D., Gable, A. L., Lyon, D., Junge, A., Wyder, S., Huerta-Cepas, J., Simonovic, M., Doncheva, N. T., Morris, J. H., Bork, P., Jensen, L. J., & Mering, C. von. (2018). STRING v11: protein–protein association networks with increased coverage, supporting functional discovery in genome-wide experimental datasets. In Nucleic Acids Research (Vol. 47, Issue D1, pp. D607–D613). Oxford University Press (OUP). doi:``` [10.1093/nar/gky1131](https://doi.org/10.1093/nar/gky1131)
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### T3DB

- **Website**: [http://www.t3db.ca](http://www.t3db.ca)
- **License**: ```T3DB is offered to the public as a freely available resource. Use and re-distribution of the data, in whole or in part, for commercial purposes requires explicit permission of the authors and explicit acknowledgment of the source material (T3DB) and the original publication (see below). We ask that users who download significant portions of the database cite the T3DB paper in any resulting publications.``` ([http://www.t3db.ca/downloads](http://www.t3db.ca/downloads))
- **Citation**:
  - ```Wishart D, Arndt D, Pon A, et al. T3DB: the toxic exposome database. Nucleic Acids Research. 2015;43(D1):D928-D934. doi:``` [10.1093/nar/gku1004](http://doi.org/10.1093/nar/gku1004)
  - ```Lim E, Pon A, Djoumbou Y, et al. T3DB: a comprehensively annotated database of common toxins and their targets. Nucleic Acids Research. 2010;38(suppl_1):D781-D786. doi:``` [10.1093/nar/gkp934](http://doi.org/10.1093/nar/gkp934)
- **Files used**:
  - `toxins.xml.zip`
  - `targets.xml.zip`
  - `toxins.csv.zip`
  - `targets.csv.zip`
  - `moas.csv.zip`
  - `structures.zip`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### TarBase

- **Website**: [http://microrna.gr/tarbase](http://microrna.gr/tarbase)
- **License**:
- **Citation**: ```Skoufos, G., Kakoulidis, P., Tastsoglou, S., Zacharopoulou, E., Kotsira, V., Miliotis, M., Mavromati, G., Grigoriadis, D., Zioga, M., Velli, A., Koutou, I., Karagkouni, D., Stavropoulos, S., Kardaras, F. S., Lifousi, A., Vavalou, E., Ovsepian, A., Skoulakis, A., Tasoulis, S. K., … Hatzigeorgiou, A. G. (2023). TarBase-v9.0 extends experimentally supported miRNA–gene interactions to cell-types and virally encoded miRNAs. In Nucleic Acids Research (Vol. 52, Issue D1, pp. D304–D310). Oxford University Press (OUP). doi:``` [10.1093/nar/gkad1071](https://doi.org/10.1093/nar/gkad1071)
- **Files used**:
  - `Homo_sapiens_TarBase_v9.tsv.gz`
  - `Mus_musculus_TarBase_v9.tsv.gz`
  - `Viral_species_TarBase-v9.tsv.gz`
  - `Other_species_TarBase_v9.tsv.gz`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### TISSUES

Tissue expression database

- **Website**: [https://tissues.jensenlab.org](https://tissues.jensenlab.org)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**: ```Palasca O, Santos A, Stolte C, Gorodkin J, Jensen LJ. TISSUES 2.0: an integrative web resource on mammalian tissue expression. Database. 2018;2018. doi:``` [10.1093/database/bay003](https://doi.org/10.1093/database/bay003)
- **Files used**:
  - `human_tissue_integrated_full.tsv`
  - `human_tissue_knowledge_full.tsv`
  - `human_tissue_experiments_full.tsv`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### TRRUST

Transcriptional Regulatory Relationships Unraveled by Sentence-based Text mining (TRRUST)

- **Website**: [https://www.grnpedia.org/trrust/](https://www.grnpedia.org/trrust/)
- **License**: CC BY-SA 4.0
- **Citation**: ```Han H, Cho JW, Lee S, et al. TRRUST v2: an expanded reference database of human and mouse transcriptional regulatory interactions. Nucleic Acids Research. 2018;46(D1):D380-D386. doi:``` [10.1093/nar/gkx1013](https://doi.org/10.1093/nar/gkx1013)
- **Files used**:
  - `trrust_rawdata.human.bioc.xml`
  - `trrust_rawdata.mouse.bioc.xml`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### TTD

- **Website**: [http://db.idrblab.net/ttd/](http://db.idrblab.net/ttd/)
- **License**:
- **Citation**: ```Zhou Y, Zhang Y, Lian X, et al. Therapeutic target database update 2022: facilitating drug discovery with enriched comparative data of targeted agents. Nucleic Acids Research. 2022;50(D1):D1398-D1407. doi:``` [10.1093/nar/gkab953](http://doi.org/10.1093/nar/gkab953)
- **Files used**:
  - `P3-01-All.sdf`
  - `P4-01-Target-KEGGpathway_all.txt`
  - `P4-06-Target-wikipathway_all.txt`
  - `P1-01-TTD_target_download.txt`
  - `P1-02-TTD_drug_download.txt`
  - `P1-03-TTD_crossmatching.txt`
  - `P1-04-Drug_synonyms.txt`
  - `P2-01-TTD_uniprot_all.txt`
  - `P1-08-Biomarker_disease.txt`
  - `P1-06-Target_disease.txt`
  - `P1-05-Drug_disease.txt`
  - `P1-09-Target_compound_activity.txt`
  - `P1-07-Drug-TargetMapping.xlsx`
  - `P2-06-TTD_sequence_all.txt`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### UberonOntology

- **Website**: [https://github.com/obophenotype/uberon](https://github.com/obophenotype/uberon)
- **License**: [CC BY 3.0](https://creativecommons.org/licenses/by/3.0/)
- **Citation**:
- **Files used**:
  - At the moment, the ontology is downloaded in OBO format (```uberon-full.obo```).

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### UNII

- **Website**: [https://fdasis.nlm.nih.gov/srs/](https://fdasis.nlm.nih.gov/srs/)
- **License**:  Except as otherwise noted, data is provided as Public Domain.
- **Citation**:
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### UniProt

- **Website**: [https://www.uniprot.org](https://www.uniprot.org)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**:
- **Files used**:

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----

### UnitOntology

- **Website**: [https://github.com/bio-ontology-research-group/unit-ontology](https://github.com/bio-ontology-research-group/unit-ontology)
- **License**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)
- **Citation**:
- **Files used**:
  - `uo-full.obo`

**Configuration properties**:

| Property       | Values     | Description                                              |
|----------------|------------|----------------------------------------------------------|
| forceExport    | true/false | Force graph export even if nothing changed               |
| ignoreObsolete | true/false | Ignore OBO ontology entries if they are flagged obsolete |

----

### USDA-PLANTS

The PLANTS Database provides standardized information about the vascular plants, mosses, liverworts, hornworts, and lichens of the U.S. and its territories.

- **Website**: [https://plants.sc.egov.usda.gov](https://plants.sc.egov.usda.gov)
- **License**: Plant information, including the distribution maps, lists, and text, is not copyrighted and is free for any use.
- **Citation**: ```USDA, NRCS. 2022. The PLANTS Database (http://plants.usda.gov, 08/09/2022). National Plant Data Team, Greensboro, NC USA.```
- **Files used**:
  -`plantlst.txt`

**Configuration properties**:

| Property    | Values     | Description                                |
|-------------|------------|--------------------------------------------|
| forceExport | true/false | Force graph export even if nothing changed |

----
